### PR TITLE
Source reducer exploration

### PR DESCRIPTION
**Pull request by hbenl**

### CODE REVIEW

**Comment by hbenl** [comment]
Fixes FE-552

**Comment by hbenl** [comment]
Fixes FE-540

File: packages/protocol/thread/thread.ts
Comment (markerikson) [inline packages/protocol/thread/thread.ts:49]: :hand: conceptually, this ends up dragging in a bunch of the "app" code.

Could we switch this up to injecting these via setter into `ThreadFront` as the app starts?

I know it's a bit silly, but it'd be nice to maintain that separation for now.

Comment (markerikson) [reply packages/protocol/thread/thread.ts:49]: That said: I've also just now realized that `ThreadFront` is used over in Brian's prototype, which has _no_ Redux store at all. So, that prototype is going to completely break because of this PR.

**Review by markerikson — COMMENTED** [review]
At least as _intended_, the changes seem okay.

**BUT**... per discussion in Discord, we've realized that this actually breaks Brian's prototype package entirely.

`packages/shared` relies on `packages/protocol`, which in turn is now both importing part of the main app's Redux store logic, and also needing a Redux store with appropriate setup to actually exist.

The import chain is an issue, but so is the implicit dependence on a valid Redux store, especially since Brian's prototype doesn't have _any_ Redux store at all.

I think it's best to split this up and try to land _just_ the `sources` reducer changes by themselves (new sources processing, selectors, etc).

Then we should figure out how to rip _all_ sources handling out of `ThreadFront` entirely and put it somewhere else.

File: src/ui/utils/sources.test.ts
Comment (jcmorrow) [inline src/ui/utils/sources.test.ts:100]: Personally, I really don't like this. `generatedFrom` should mean that we have a sourcemap linking source `pp1` with the generated source `1`, and that is how we computed the content of `pp1`. Generated and original sources only make sense in the context of source-mapping, and that is not what pretty-printing is (or, if it was, it would be correct to call the pretty-printed file the `generated`, not the `original`). I'm not going to block on it, because I am not doing much with this code these days, but I find this a major source of confusion. I think if we continue to completely reverse the meanings of these terms in some contexts we will also continue to completely confuse the daylight out of every single person who ever tries to understand the code.

File: src/ui/utils/sources.ts
Comment (jcmorrow) [inline src/ui/utils/sources.ts:34]: This code is incredibly confusing, even for me, and I wrote it. I think it's fine if we rework it to be more efficient, but not at the cost of removing all explanations of how it works. I would like to understand:

1) Are these changes more performant? How much more?
2) What did we change about how we process sources?
3) Can we leave some explanation about the trickier parts of this code? It can live in the commit, or in separate documentation, or inline, but I think future us will need it somewhere, at some point.

File: src/ui/utils/sources.ts
Comment (jcmorrow) [inline src/ui/utils/sources.ts:32]: If we don't need to do the corresponding check I think it would be easier to read this as
`const newSourcesById = groupBy(newSources, "sourceId)` but I don't feel strongly about it.

Comment (jcmorrow) [reply src/ui/utils/sources.ts:32]: Failing making that change we should at least have line 30 be `for (const ...` rather than `for (let ...`

File: packages/protocol/thread/thread.ts
Comment (hbenl) [reply packages/protocol/thread/thread.ts]: > Could we switch this up to injecting these via setter into ThreadFront as the app starts?

Done.

File: src/ui/utils/sources.test.ts
Comment (hbenl) [reply src/ui/utils/sources.test.ts:100]: We have 3 different reasons for original <-> generated links between sources: sourcemaps, pretty-printing and inline script extraction. We could store these 3 types of links separately but in our code we mostly use these links to determine which source is preferred for display - separating the links would make these parts of our code (slightly) more complicated while I can't think of a single place where it would help make the code simpler.
If we want to avoid confusion I think we should use different terms for these links but not make working with them in the code harder.

File: src/ui/utils/sources.ts
Comment (hbenl) [reply src/ui/utils/sources.ts:32]: `groupBy` would create a `Dictionary<newSource[]>`, but we know that there is only one source per sourceId, so I prefer not using that (though I don't feel strongly about it either).
I'll replace all `for (let ...` with `for (const ...`.

File: src/ui/utils/sources.ts
Comment (hbenl) [reply src/ui/utils/sources.ts:34]: 1. I didn't rewrite this to be more performant but to be less confusing. I think it's rather straightforward now.
2. no more `canonicalId` and the links between pretty-printed and minified sources are not removed from `generated`/`generatedFrom` anymore
3. I've added a few comments and could add more if you tell me which parts you find confusing

File: src/devtools/client/webconsole/components/Search/useConsoleSearch.test.tsx
Comment (markerikson) [inline src/devtools/client/webconsole/components/Search/useConsoleSearch.test.tsx:92]: :question: How did this snapshot change?

**Review by markerikson — APPROVED** [review]
Looks pretty good overall!

File: src/devtools/client/webconsole/components/Search/useConsoleSearch.test.tsx
Comment (hbenl) [reply src/devtools/client/webconsole/components/Search/useConsoleSearch.test.tsx:92]: The vscode jest extension offered to update this "inline snapshot".
The test and the hook that it tests are a complete mystery to me, far too much React magic. I could only guess from the name that it has to do with searching in the replay console so I did some manual tests and didn't notice any differences in behavior.
@bvaughn could you have a look if this snapshot update makes sense?

Comment (bvaughn) [reply src/devtools/client/webconsole/components/Search/useConsoleSearch.test.tsx:92]: I'd be happy to look, but this doesn't feel like a very good hand off. Maybe next time we can verify that an unexpected change to snapshot values wasn't a regression before we land the change.

Comment (bvaughn) [reply src/devtools/client/webconsole/components/Search/useConsoleSearch.test.tsx:92]: Just verified that the committed search result ids still look correct, given the message contents. (The same messages are still being matched, just with slightly different ids).

I think something about this change impacted the ordering of, or id generation for, "newSource" messages in the Redux store. (That's all the ids are.)

**Pull request merged**
